<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [67]:
import pandas as pd
import psycopg2
import warnings
warnings.filterwarnings("ignore")

In [69]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [70]:
# текст запроса
query_3_1 = f'''
select count(id) cnt_vacancies --считаем количество id (уникальный номер вакансии) в таблице с вакансиями
from VACANCIES --таблица вакансии
'''

In [71]:
# результат запроса
#записываем в DF результаты запроса 
df_3_1 = pd.read_sql_query(query_3_1, connection) 
#выводим DF на экран
display(df_3_1) 

cnt_vacancies
0          49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [72]:
# текст запроса
query_3_2 = f'''
select count(id) cnt_employers --считаем количество id (уникальный номер работодателя) в таблице с работодателями
from EMPLOYERS --таблица работодатели
'''

In [73]:
# результат запроса
#записываем в DF результаты запроса 
df_3_2 = pd.read_sql_query(query_3_2, connection) 
#выводим DF на экран
display(df_3_2) 

cnt_employers
0          23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [74]:
# текст запроса
query_3_3 = f'''
select count(id) cnt_areas --считаем количество id (уникальный номер региона) в таблице с регионами
from AREAS --таблица регионы
'''

In [75]:
# результат запроса
#записываем в DF результаты запроса 
df_3_3 = pd.read_sql_query(query_3_3, connection) 
#выводим DF на экран
display(df_3_3) 

cnt_areas
0       1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [76]:
# текст запроса
query_3_4 = f'''
select count(id) cnt_industries --считаем количество id (уникальный номер сферы) в таблице со сферами деятельности
from INDUSTRIES --таблица сферы деятельности
'''

In [77]:
# результат запроса
#записываем в DF результаты запроса 
df_3_4 = pd.read_sql_query(query_3_4, connection) 
#выводим DF на экран
display(df_3_4) 

cnt_industries
0             294

***

# выводы по предварительному анализу данных
В базе данных хранятся данные по 49197 вакансиям, 23501 работодателям, 1362 регионам и 294 сферам деятельности.

В БД есть 6 таблиц, из которых 1 основная - Вакансии, 3 справочника (по работодателю, региону и сфере деятельности), также есть дополнительная таблица для организации связи между таблицей с работодателями и сферами деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [78]:
# текст запроса
query_4_1 = f'''
select
  count(v.id) cnt_vacancies, --считаем количество id (уникальный номер вакансии)
  a.name area --название региона из связанной таблицы-справочника
from VACANCIES v --таблица с вакансиями
join AREAS a on v.area_id = a.id --добавляем данные из справочника с регионами по ключу с id  региона 
group by a.name --группируем по названию региона
order by count(v.id) desc --сортируем по количеству в порядке убывания
--limit 5 --выводим ТОП 5
'''

In [79]:
# результат запроса
#записываем в DF результаты запроса 
df_4_1 = pd.read_sql_query(query_4_1, connection) 
#выводим DF на экран
display(df_4_1) 

cnt_vacancies             area
0             5333           Москва
1             2851  Санкт-Петербург
2             2112            Минск
3             2006      Новосибирск
4             1892           Алматы
..             ...              ...
764              1       Тарко-Сале
765              1    Новоаннинский
766              1            Бирск
767              1           Сасово
768              1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [80]:
# текст запроса
query_4_2 = f'''
select
  count(id) cnt_with_salary --считаем количество id (уникальный номер вакансии)
from VACANCIES --таблица с вакансиями
where (salary_from is not null) or (salary_to is not null) --фильтруем по вакансиям, где указана хотя бы одна из ЗП
'''

In [81]:
# результат запроса
#записываем в DF результаты запроса 
df_4_2 = pd.read_sql_query(query_4_2, connection) 
#выводим DF на экран
display(df_4_2) 

cnt_with_salary
0            24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [82]:
# текст запроса
query_4_3 = f'''
select
  round(avg(salary_from)) avg_salaryfrom, --среднее значение нижней границы зарплаты
  round(avg(salary_to)) avg_salaryto --среднее значение верхней границы зарплаты
from VACANCIES --таблица с вакансиями
'''

In [83]:
# результат запроса
#записываем в DF результаты запроса 
df_4_3 = pd.read_sql_query(query_4_3, connection) 
#выводим DF на экран
display(df_4_3) 

avg_salaryfrom  avg_salaryto
0         71065.0      110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [84]:
# текст запроса
query_4_4 = f'''
select
  count(id) cnt_vacancies, --считаем количество id (уникальный номер вакансии)
  schedule, --тип рабочего графика
  employment --тип трудоустройства
from VACANCIES --таблица с вакансиями
group by schedule, employment --группируем по типу рабочего графика и типу трудоустройства
order by count(id) desc --сортируем по количеству в порядке убывания
--offset 1 limit 1 --выводим вторую строку (не учитываем первую, после выводим одну строку)
'''

In [85]:
# результат запроса
#записываем в DF результаты запроса 
df_4_4 = pd.read_sql_query(query_4_4, connection) 
#выводим DF на экран
display(df_4_4) 

cnt_vacancies          schedule           employment
0           35367       Полный день     Полная занятость
1            7802  Удаленная работа     Полная занятость
2            1593     Гибкий график     Полная занятость
3            1312  Удаленная работа  Частичная занятость
4             940    Сменный график     Полная занятость
5             569       Полный день           Стажировка
6             367    Вахтовый метод     Полная занятость
7             347       Полный день  Частичная занятость
8             312     Гибкий график  Частичная занятость
9             141       Полный день     Проектная работа
10            133  Удаленная работа     Проектная работа
11            116     Гибкий график           Стажировка
12            101    Сменный график  Частичная занятость
13             64  Удаленная работа           Стажировка
14             18     Гибкий график     Проектная работа
15             12    Сменный график           Стажировка
16              2    Вахтовый метод     Проектная работа
17              1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [86]:
# текст запроса
query_4_5 = f'''
select
  experience --требуемый опыт работы
from VACANCIES --таблица с вакансиями
group by experience --группируем по опыту работы
order by count(id) --сортируем по количеству вакансий в порядке возрастания
'''

In [87]:
# результат запроса
#записываем в DF результаты запроса 
df_4_5 = pd.read_sql_query(query_4_5, connection) 
#выводим DF на экран
display(df_4_5) 

experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

# выводы по детальному анализу вакансий
Наибольшее количество вакансий сосредоточено в крупных городах: Москва (5333), Санкт-Петербург (2851), Минск (2112), Новосибирск (2006), Алматы (1892).

Зарплатная вилка от 71065 до 110537, при этом уровень дохода указан только для 24073 вакансий (примерно половина от общего числа).

Наибольшее количество вакансий расчитано на полную занятость с различным графиком: полный день - 35367, удаленная работа - 7802, гибкий график - 1593.

Чаще всего требуется опыт от 1 года до 3 лет, реже всего ищут специалистов с опытом более 6 лет.


Таким образом, в каждом из запросов есть группа, которая выделяется среди остальных. По этой причине, при обработке данных, возможно, есть смысл выделить ключевые группы, а остальные объединить.

В случае с городами - выделяем вакансии городов миллионников, остальные вакансии объединяем либо в одну группу "Остальные", либо разделяем по территориальному признаку (например, центральная Россия, Урал и т.д.). Необходим дополнительный анализ.

При анализе занятости-графика были выявлены 3 ключевые группы: полная занятость/полный день, полная занятость/удаленная работа, полная занятость/гибкий график. Остальные также можно объединить либо в одну, либо в несколько ключевых групп. Также необходим доп. анализ.

Зарплатная вилка скорей всего сильно отличается в зависимости от региона, требуется доп. анализ.

Чаще всего ищут специалистов с опытом работы от 1 года до 3 лет. Скорей всего это обусловлено тем, что работодателю важно, чтобы у сотрудника был некоторый опыт, но при этом была возможность платить меньшую ЗП (из расчета, что сотрудники с большим опытом рассчитывают на более высокую ЗП).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [88]:
# текст запроса
query_5_1 = f'''
(select
  count(v.id) cnt_vacancies, --считаем количество id (уникальный номер вакансии)
  e.name employer --название работодателя
from EMPLOYERS e  --таблица с работодателями
left join VACANCIES v on v.employer_id = e.id --добавляем данные из таблицы с вакансиями по ключу с id  работодателя 
group by e.id, e.name --группируем по работодателю
order by count(v.id) desc --сортируем по количеству в порядке убывания
limit 1 --выводим первую строку
)
union --объединение
(select
  count(v.id), --считаем количество id (уникальный номер вакансии)
  e.name --название работодателя
from EMPLOYERS e --таблица с работодателями
left join VACANCIES v on v.employer_id = e.id --добавляем данные из таблицы с вакансиями по ключу с id  работодателя 
group by e.id, e.name --группируем по работодателю
order by count(v.id) desc --сортируем по количеству в порядке убывания
offset 4 limit 1 --не учитываем первые 4 сроки, после выводим одну строку
)
order by 1 desc -- сортируем по убыванию количества
'''

In [89]:
# результат запроса
#записываем в DF результаты запроса 
df_5_1 = pd.read_sql_query(query_5_1, connection) 
#выводим DF на экран
display(df_5_1) 

cnt_vacancies       employer
0           1933         Яндекс
1            331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [90]:
# текст запроса
query_5_2 = f'''
select
a.name area, --название региона
count(e.id) cnt_employers --количество работодателей
from AREAS a --таблица регионы
left join VACANCIES v on v.area_id = a.id --добавляем данные из таблицы с вакансиями по ключу с id региона
left join EMPLOYERS e on e.area = a.id --добавляем данные из таблицы с работодателями по ключу с id региона
where v.id is null --фильтруемся по строкам, где нет вакансий
group by a.id, a.name --группируем по региону
order by count(e.id) desc --группируем по количетсву работодателей
'''

In [91]:
# результат запроса
#записываем в DF результаты запроса 
df_5_2 = pd.read_sql_query(query_5_2, connection) 
#выводим DF на экран
display(df_5_2) 

area  cnt_employers
0                Россия            410
1             Казахстан            207
2    Московская область             75
3    Краснодарский край             19
4    Ростовская область             18
..                  ...            ...
588             Скидель              0
589          Белоозерск              0
590           Орловский              0
591          Павловская              0
592             Таксимо              0

[593 rows x 2 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [92]:
# текст запроса
query_5_3 = f'''
select
e.name employer, --название работодателя
count(distinct v.area_id) cnt_areas --количество уникальных регионов с вакансиями работодателя
from EMPLOYERS e --таблица работодатели
left join VACANCIES v on v.employer_id = e.id --добавляем данные из таблицы с вакансиями по ключу с id региона
group by e.id, e.name --группируем по работодателю
order by count(v.area_id) desc --сортируем по количеству в порядке возрастания
'''

In [93]:
# результат запроса
#записываем в DF результаты запроса 
df_5_3 = pd.read_sql_query(query_5_3, connection) 
#выводим DF на экран
display(df_5_3) 

employer  cnt_areas
0              Яндекс        181
1          Ростелеком        152
2            Тинькофф         43
3                СБЕР         24
4       Газпром нефть         24
...               ...        ...
23496          МАКМАХ          0
23497   Завод Стелкон          0
23498        ИнтерЭВМ          0
23499  Фёдоров Михаил          0
23500      ИнтерТраст          0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [94]:
# текст запроса
query_5_4 = f'''
select
count(e.id) cnt_employers --количество работодателей
from EMPLOYERS e --таблица работодатели
left join EMPLOYERS_INDUSTRIES i on i.employer_id = e.id --добавляем данные из доп. таблицы-связки со сферами деятельности по ключу с id работодателя
where i.industry_id is null --фильтруем по id сферы (оставляем те, где не указана)
'''

In [95]:
# результат запроса
#записываем в DF результаты запроса 
df_5_4 = pd.read_sql_query(query_5_4, connection) 
#выводим DF на экран
display(df_5_4) 

cnt_employers
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [96]:
# текст запроса
query_5_5 = f'''
select
e.name employer_4ind --название работодателя
from EMPLOYERS e --таблица работодатели
left join EMPLOYERS_INDUSTRIES ei on ei.employer_id = e.id --добавляем данные из доп. таблицы-связки со сферами деятельности по ключу с id работодателя
group by e.id, e.name
having count(ei.industry_id) = 4
order by e.name
'''

In [97]:
# результат запроса
#записываем в DF результаты запроса 
df_5_5 = pd.read_sql_query(query_5_5, connection) 
#выводим DF на экран
display(df_5_5) 

employer_4ind
0        101 Интернет
1            21vek.by
2                2ГИС
3                  2К
4         4 пикселя +
...               ...
1133            ЮРИОН
1134  ЮТИП Технологии
1135           ЯКласс
1136            ЯрНео
1137    Яшин&Партнёры

[1138 rows x 1 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [98]:
# текст запроса
query_5_6 = f'''
select
count(ei.employer_id) cnt_employers --количество работодателей
from EMPLOYERS_INDUSTRIES ei --таблица работодатели
left join INDUSTRIES i on ei.industry_id = i.id --добавляем данные из доп. таблицы-связки со сферами деятельности по ключу с id работодателя
where i.name = 'Разработка программного обеспечения' --фильтруем по нужной сфере деятельности
'''

In [99]:
# результат запроса
#записываем в DF результаты запроса 
df_5_6 = pd.read_sql_query(query_5_6, connection) 
#выводим DF на экран
display(df_5_6)

cnt_employers
0           3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [100]:
# код для получения списка городов-милионников
#импорт необходимых библиотек
import requests
from bs4 import BeautifulSoup

#выполняем GET-запрос, содержимое ответа записываем в переменную response
response = requests.get('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
#создаём объект BeautifulSoup, указывая html-парсер
page = BeautifulSoup(response.text, 'html.parser')
#получаем список ссылок с городами по тегу и соответствующему классу
links = page.findAll('div', class_='ts-ПозКарта-метка')
#создаем список для названий городов-миллионников
cities = []
#цикл для получения из ссылок текста с названием города
for i in links:
    #добавляем в список городов текст из ссылки с названием города
    cities.append(i.text)
#преобразуем в кортеж (для использования в sql-запросе)   
cities = tuple(cities)
#вывод городов для проверки 
print(cities)  

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [101]:
# текст запроса
query_5_7 = f'''
(select
a.name area, --название региона
count(v.id) cnt_vacancies --количество вакансий
from  VACANCIES v --таблица вакансии
join AREAS a on v.area_id = a.id --добавляем данные из таблицы справочника регионов по ключу с id региона
join EMPLOYERS e on v.employer_id = e.id --добавляем данные из таблицы справочника работодателей по ключу с id работодателя
where e.name = 'Яндекс' and a.name in {cities} --фильтруем вакансии: оставляем только те, где работодатель Яндекс и регион из списка городов-миллионников
group by a.name) --группируем по названию региона
union --присоединяем строку total
(select
'Total', --обозначение строки Total
count(v.id) --количество вакансий
from  VACANCIES v --таблица вакансии
join AREAS a on v.area_id = a.id --добавляем данные из таблицы справочника регионов по ключу с id региона
join EMPLOYERS e on v.employer_id = e.id --добавляем данные из таблицы справочника работодателей по ключу с id работодателя
where e.name = 'Яндекс' and a.name in {cities}) --фильтруем вакансии: оставляем только те, где работодатель Яндекс и регион из списка городов-миллионников
order by 2 -- сортируем по возрастанию количества вакансий
'''

In [102]:
# результат запроса
#записываем в DF результаты запроса 
df_5_7 = pd.read_sql_query(query_5_7, connection) 
#выводим DF на экран
display(df_5_7)

area  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

# выводы по анализу работодателей
Наибольшее количество вакансий у Яндекса (1933), который охватывает больше всего регионов (181). При этом при анализе городов-миллионников наибольшее количество вакансий Яндекса в Москве (54), далее Санкт-Петербург (43), замыкает тройку Екатеринбург (39).
Больше всего компаний, в которых нет вакансий, из России (410), далее идет Казахстан (207).
По количеству регионов с вакансиями после Яндекса идут Ростелеком (152) и Тинькофф (43), далее Сбер и Газпромнефть (по 24).
В сфере разработки ПО задействованы 3553 работодателей (примерно шестая часть от всех работодателей).

Таким образом,Яндекс охватывает наибольшее количество регионов при наибольшем количестве вакансий. Возможно, это связано с возможностью удаленной работы. Необходимо дополнительно проанализировать вакансии компании в разрезе занятости-графика.
Для анализа нет смысла рассматривать компании, в которых нет вакансий. Их не так много относительно общего количества работодателей.
Разработка ПО - основной маркер IT-компаний, т.е. потенциальных работодателей для DS-специалистов. Однако направление DS может развиваться не только в компаниях, связанных с IT. Необходим дополнительный анализ - в каких сферах могут требоваться DS-специалисты.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [103]:
# текст запроса
query_6_1 = f'''
select
  count(v.id) cnt_vacanties --считаем количество id (уникальный номер вакансии)
from VACANCIES v --таблица с вакансиями
where lower(v.name) like '%data%' or lower(v.name) like '%данн%' --фильтр названию вакансии (в нижнем регистре), где встручаются слова 'data' или 'данн'
'''

In [104]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection) 
#выводим DF на экран
display(df_6_1)

cnt_vacanties
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [105]:
# текст запроса
query_6_2 = f'''
select
  count(v.id) cnt_vacancies --считаем количество id (уникальный номер вакансии)
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
and (lower(v.name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка') --фильтр по уровню специалиста
'''

In [106]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection) 
#выводим DF на экран
display(df_6_2)

cnt_vacancies
0             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [107]:
# текст запроса
query_6_3 = f'''
select
  count(v.id) cnt_vacancies --считаем количество id (уникальный номер вакансии)
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
and (lower(key_skills) similar to '%sql%|%postgres%') --фильтр по ключевому навыку
'''

In [108]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection) 
#выводим DF на экран
display(df_6_3)

cnt_vacancies
0            229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [109]:
# текст запроса
query_6_4 = f'''
select
  count(v.id) cnt_vacancies --считаем количество id (уникальный номер вакансии)
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
and (lower(v.key_skills) like '%python%') --фильтр по ключевому навыку
'''

In [110]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection) 
#выводим DF на экран
display(df_6_4)

cnt_vacancies
0            357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [111]:
# текст запроса
query_6_5 = f'''
select
round(avg(length(v.key_skills) - (length(replace(v.key_skills, chr(9), '')) - 1)), 2) cnt_key_skills --вычисляем количество навыков: длина строки - длина строки без разделителей - 1. Далее находим целое и округляем до 2 знаков
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (v.name like '%ML%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
'''

In [112]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection) 
#выводим DF на экран
display(df_6_5)

cnt_key_skills
0            6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [113]:
# текст запроса
query_6_6 = f'''
--промежуточная таблица для вычисления средней ЗП в зависимости от указанной вилки
with x as 
(
select
experience, --опыт работы
case when ((salary_from is not null) and (salary_to is not null)) then ((salary_from + salary_to)/2) --если оба значения не пустые, берем среднее
else coalesce(v.salary_from, v.salary_to) --иначе берем то, которое не пустое (строки, где оба значения пустые отсееваем фильтром ниже)
end salary
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (v.name like '%ML%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
and
((salary_from is not null) and (salary_to is not null)) --фильтр, если не указано ни одно значение в вилке
)
--получаем данные из промежуточной таблицы
select
experience, --опыт работы
round(avg(salary)) avg_salary --среднее значение ЗП, округленное до целого
from x --промежуточная таблица
group by experience --группируем по опыту работы
'''

In [114]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection) 
#выводим DF на экран
display(df_6_6)

experience  avg_salary
0       От 3 до 6 лет    248830.0
1           Нет опыта     89167.0
2  От 1 года до 3 лет    162182.0

***

# выводы по предметному анализу
Всего в базе 1771 вакансия по направлению DS. Из них для начинающего специалиста подойдут 51 вакансия.

Навык SQL требуется в 229 вакансиях, а Python в 357.

В среднем в вакансии требуется 6-7 навыков.

Зарплата в среднем для начинающего специалиста составляет 74643, для специалиста с опытом от 1 года до 3 лет - 139675, для опыта от 3 до 6 лет - 243115.


Таким образом, для более детального анализа необходимо выделить ТОП навыков, которые требуются от соискателя по DS-направлению. 

Благодаря этому можно будет понять, какие навыки являются самыми востребованными на рынке, а значит наиболее значимыми.

Зарплата, как говорилось ранее, скорей всего будет отличаться в зависимости от региона, поэтому для полноты картины можно дополнительно изучить ее в разрезе регионов.

# Общий вывод по проекту
подведем итог исследования, обобщите выводы
десь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Таким образом, анализ показал, что наиболее востребованны специалисты, готовые работать полный рабочий день по полной занятости с опытом работы от 1 года до 3 лет.
Компания Яндекс охватывает наибольшее количество регионов и предлагает максимум вакансий.
Средняя ЗП для DS-специалиста с опытом работы превышает общую зарплатную вилку.

В выводах для каждого блока были выдвинуты некоторые предложения относительно дополнительных исследований. Рассмотрим некоторые из них.

1. Рассмотрим изменение ЗП вилки в зависимости от региона.

In [115]:
# текст запроса
query = f'''
select
  a.name, --название региона
  round(avg(v.salary_from)) avg_salaryfrom, --среднее значение нижней границы зарплаты
  round(avg(v.salary_to)) avg_salaryto --среднее значение верхней границы зарплаты
from VACANCIES v --таблица с вакансиями
join AREAS a on v.area_id = a.id --соединяем с таблицей регионов по id региона
where v.salary_from is not null and v.salary_to is not null --фильтр по заполненной ЗП
group by a.id, a.name --группируем по региону
order by 2 desc --сортируем по ЗП
'''

In [116]:
# результат запроса
df = pd.read_sql_query(query, connection) 
#выводим DF на экран
display(df)

name  avg_salaryfrom  avg_salaryto
0       Германия        348270.0      557232.0
1     Черногория        282501.0      331208.0
2    Таджикистан        233794.0      292242.0
3        Кишинёв        233794.0      292242.0
4         Сербия        219182.0      277631.0
..           ...             ...           ...
471       Алушта         12534.0       25000.0
472        Семей         12205.0       18307.0
473        Тараз         12205.0       15459.0
474           Ош          7282.0       18204.0
475        Тосно          6000.0        6200.0

[476 rows x 3 columns]

Таким образом, наиболее высокие ЗП в заграничных вакансиях, наименьшие - в мальких провинциальных городах. Т.е. подтверждаем гипотезу о том, что регион довольно сильно влияет на ЗП.

2. Проверяем гипотезу о том, что Яндекс представлен во многих регионах за счет возможности удаленной работы.

In [117]:
# текст запроса
query = f'''
select
  count(v.id) cnt_vacancies, --считаем количество id (уникальный номер вакансии)
  count(distinct v.area_id), --считаем количество регионов
  v.schedule, --тип рабочего графика
  v.employment --тип трудоустройства
from VACANCIES v --таблица с вакансиями
join EMPLOYERS e on e.id = v.employer_id
where e.name = 'Яндекс' --фильтр по компании
group by v.schedule, v.employment --группируем по типу рабочего графика и типу трудоустройства
order by count(v.id) desc --сортируем по количеству в порядке убывания
'''

In [118]:
# результат запроса
df = pd.read_sql_query(query, connection) 
#выводим DF на экран
display(df)

cnt_vacancies  count          schedule           employment
0           1045    175  Удаленная работа  Частичная занятость
1            645    173  Удаленная работа     Полная занятость
2            125     27       Полный день     Полная занятость
3             55     34    Сменный график  Частичная занятость
4             35     20     Гибкий график  Частичная занятость
5             14      5     Гибкий график     Полная занятость
6              6      3       Полный день           Стажировка
7              6      5     Гибкий график           Стажировка
8              2      2    Сменный график     Полная занятость

Таким образом, наибольшее количество вакансий Яндекса рассчитано на удаленную работу, а так как Яндекс является одной из самых крупных IT-компаний, можем сделать вывод, что скорей всего большое количество вакансий в сфере IT могут подразумевать возможность удаленной работы, т.е. регион уже имеет не такой большой вес для соискателя.

3. Посмотрим какой разброс средней ЗП в вакансиях в сфере DS в зависимости от опыта работы.

In [119]:
query = f'''
select
experience, --опыт работы
case when ((salary_from is not null) and (salary_to is not null)) then ((salary_from + salary_to)/2) --если оба значения не пустые, берем среднее
else coalesce(v.salary_from, v.salary_to) --иначе берем то, которое не пустое (строки, где оба значения пустые отсееваем фильтром ниже)
end salary
from VACANCIES v --таблица с вакансиями
where
(lower(v.name) similar to '%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%'
or (v.name like '%ML%' and lower(v.name) not like '%html%')) --фильтр по названию вакансии для DS: проверяем похоже ли название на что-то из перечня, отдельно проверяем похоже ли на ML, при этом исключая HTML
and
((salary_from is not null) or (salary_to is not null)) --фильтр, если не указано ни одно значение в вилке
'''

In [120]:
# результат запроса
df = pd.read_sql_query(query, connection) 
#выводим DF на экран
display(df)

experience  salary
0   От 1 года до 3 лет   60000
1   От 1 года до 3 лет   75000
2   От 1 года до 3 лет   60000
3            Нет опыта   37000
4            Нет опыта   50000
..                 ...     ...
57  От 1 года до 3 лет  100000
58       От 3 до 6 лет  292242
59  От 1 года до 3 лет  175000
60  От 1 года до 3 лет  143500
61       От 3 до 6 лет   67126

[62 rows x 2 columns]

Для наглядности построим график.

In [121]:
#импорт необходимых библиотек
import plotly.express as px
from plotly.subplots import make_subplots

#добавляем коробчатую диаграмму
fig = px.box(data_frame=df, y='salary', x='experience')
#задаем доп. параметры (заголовок, отсутствие легенды, подпись осей)
fig.update_layout(title="Распределение уровня заработной платы в зависимости от требуемого опыта работы",
                  showlegend=False,
                  xaxis_title="Опыт",
                  yaxis_title="ЗП (руб)")

#сохранение графика
fig.write_html("plotly/graph.html")

#выводим график
fig.show()

Из графика видно, что для специалист с опытом может рассчитывать на ЗП от 67 до 450 тыс руб. (в среднем 233 тыс. руб.), с опытом от 1 года до 3 лет - от 50 до 275 тыс. руб. (в среднем 130 тыс. руб.), и специалист без опыта может получать оти 37 до 115 тыс. руб. (в среднем 72 тыс. руб.).
Таким образом, при дальнейшей работе надо моделью важно учитывать соответствие ЗП вилки в вакансиях с зарплатными ожиданиями соискателя.

В качеастве продолжения работы можно проверить оставшиеся гипотезы, выраженные в выводах по блокам. Например, выделить основные навыки, которые требуются для DS специалистов или провести анализ - все ли вакансии, требующие навыки IT от компаний из IT-сферы.

In [122]:
#Закрываем соединение
connection.close()